### PR TITLE
Run-time types hooks

### LINKED ISSUE #1215: Strings not freed on runtime components
**Describe the bug**

When creating runtime components that contain strings, these strings are not freed on entity destruction, resulting in a memory leak.

**To Reproduce**

This snippet sets up a hook on the `free()` function used by Flecs and prints a message if the pointer to the string we are expecting to be deleted is actually deleted. This deletion never happens, thus there is a memory leak.

```cpp

#include <flecs.h>

ecs_os_api_free_t oldfree;

void *strptr = nullptr; // this will point to the string we are expecting to be freed.

void myfree(void *ptr) {
  if (ptr) {
    if (strptr && strptr == ptr) {
     // print a message if the pointer we are looking for is to be deleted.
      std::cout << "gotcha" << std::endl;
    }
  }
  oldfree(ptr);
}

void main() {

  oldfree = ecs_os_api.free_; // set up a hook on the free() function used by Flecs
  ecs_os_api.free_ = myfree;

  // create a runtime component with one string member
  auto comp = ecs.component("mycomp");
  comp.member(flecs::String, "text");

  flecs::entity e = ecs.entity();
  e.add(comp);

  void *ptr = e.get_mut(comp);

  flecs::cursor cur(ecs, comp, ptr);

  cur.push();
  cur.member("text");
  cur.set_string("hola");
  // get a pointer to the string stored in the component instance:
  uintptr_t *s = (uintptr_t *) cur.get_ptr(); 
  strptr = (void *) (*s);

  cur.pop();

  std::cout << "about to destruct" << std::endl;
  e.destruct(); // at some point here we should expect the string to be destroyed and "gotcha" should be displayed, but it is not.
  std::cout << "after destruct" << std::endl;

}

// output:
// about to destruct
// after destruct
```

I also could not find in the code (`addons/meta`) anything that automatically frees up resources related to runtime components.

**Expected behavior**

When an entity that contains a runtime component is destroyed, all items in that runtime struct should be destroyed.

### PR DESCRIPTION
**Pull request by jpeletier**
This PR extends Flecs' runtime type support by providing automatically generated constructors/destructors and move/copy hooks. This means that non-trivial runtime types that have a mix of primitive and/or complex types/CPP types can now be instantiated/destroyed / moved (as with an archetype change) or copied (as when a prefab is instantiated) correctly, without memory leaks.

It also enables arrays runtime support for arrays so that arrays of complex types, which can either be native types or other runtime types are also correctly instantiated, destroyed, etc.

Fixes #1215 
Reverts #1309, since a world pointer is no longer necessary in `ecs_type_info_t`

Example:
```cpp
  struct A {
    int value = 7;
    A() { printf("Default constructed A, value=%d\n", value); }

    A &operator=(const A &other) {
      value = other.value;
      printf("Copy assigned A, value=%d\n", value);
      return *this;
    }

    A &operator=(A &&other) {
      value = other.value;
      other.value = 0;
      printf("Move assigned A, value=%d\n", value);
      return *this;
    }
    ~A() { printf("Destroyed A, value=%d\n", value); }
  };

  ecs.component<A>("TheA").member<int>("value");

  flecs::entity rtt = ecs.component("RuntimeType").member<A>("a");
  flecs::entity prefab = ecs.prefab();
  void *ptr = prefab.ensure(rtt);
  // output:
  //   Default constructed A, value=7

  // set prefab's rtt .a.value to 100 using a cursor
  flecs::cursor cur(ecs, rtt, ptr);
  cur.push();
  cur.member("a");
  cur.push();
  cur.member("value");
  cur.set_int(100);

  // Instantiate the prefab:
  flecs::entity e = ecs.entity("E").is_a(prefab);
  // output:
  //  Default constructed A, value=7
  //  Copy assigned A, value=100

  // add another entity to e, which will trigger an archetype change and thus a move:
  e.add(ecs.entity());
  // output:
  //   Default constructed A, value=7
  //   Move assigned A, value=100
  //   Destroyed A, value=0

  e.destruct();
  // output:
  //   Destroyed A, value=100

  prefab.destruct();
  // output:
  //   Destroyed A, value=100
```

### CODE REVIEW

File: test/script/src/Eval.c [hunk 1]
```diff
@@ -4063,8 +4063,6 @@ void Eval_multiline_string(void) {
         "Special characters }{\"\"'',\n"
     );
 
-    ecs_os_free(ptr->value);
```
Comment (jpeletier) [inline test/script/src/Eval.c:4066]: Notice that this is no longer required, and actually segfaults now, since the string is freed by the generated destructor for the runtime struct defined in this test

**Review by SanderMertens — CHANGES_REQUESTED** [review]

File: test/script/src/Eval.c (hunk 1 above)
Comment (SanderMertens) [reply test/script/src/Eval.c:4066]: LGTM

File: src/addons/meta/runtime.c [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+  bool move_hook_required = array_ti->hooks.move != NULL;
+  bool copy_hook_required = array_ti->hooks.copy != NULL;
+
+  ecs_meta_rtt_ctx_t *rtt_ctx = flecs_meta_rtt_configure_hooks(world, ti, ctor_hook_required, dtor_hook_required,
+                                                               move_hook_required, copy_hook_required);
+
+  if (!rtt_ctx)
+    return;  // no hooks required
+
+  if (ctor_hook_required) {
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c:257]: I am not sure I understand what you mean. By the first if-else-if statement in your suggestion you're implying that there will be either ctor or dtor but not both.
Also, each hook must have its own entry. Even if they both have the same type info, offset and count, I am sharing the context type `ecs_meta_rtt_ctx_t` and call data information `ecs_meta_rtt_call_data_t` type for both structs and arrays.
In a future iteration, we could have a `ecs_meta_rtt_ctx_t` that is specific for arrays, as an optimization. This would have a common count/offset/type_info at the context level, saving memory and make that code shorter as you want.

File: src/addons/meta/runtime.c [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+  bool ctor_hook_required = false;
+  bool dtor_hook_required = false;
+  bool move_hook_required = false;
+  bool copy_hook_required = false;
+
+  // Iterate all struct members and see if hooks are required:
+  for (int i = 0; i < ecs_vec_count(&struct_info->members); i++) {
+    ecs_member_t *m = ecs_vec_get_t(&struct_info->members, ecs_member_t, i);
+    const ecs_type_info_t *member_ti = ecs_get_type_info(world, m->type);
+    ctor_hook_required |= member_ti->hooks.ctor && member_ti->hooks.ctor != flecs_default_ctor;
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c:175]: I think so, because I don't know who set that `flecs_default_ctor` constructor, perhaps even the user did in a custom component.

File: src/addons/meta/runtime.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (SanderMertens) [reply src/addons/meta/runtime.c:257]: My bad, I've updated my example to fix the issue. It's just a suggestion to reduce the amount of code duplication.

File: src/addons/meta/runtime.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (SanderMertens) [reply src/addons/meta/runtime.c:175]: If the user explicitly provided `flecs_default_ctor` it should not be overridden. This should be no different than a user providing a custom function that initializes to 0. If possible I'd prefer it if there are no special semantics for `flecs_default_ctor`.

File: src/addons/meta/runtime.c [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+  if (!rtt_ctx)
+    return;  // no hooks required
+
+  // At least a hook was configured, therefore examine each struct member to build the vector of calls that will then
+  // be executed by the generic hook handler:
+  for (int i = 0; i < ecs_vec_count(&struct_info->members); i++) {
+    ecs_member_t *m = ecs_vec_get_t(&struct_info->members, ecs_member_t, i);
+    const ecs_type_info_t *member_ti = ecs_get_type_info(world, m->type);
+    if (ctor_hook_required) {
```
[the remark this answers is not in the document]
Comment (SanderMertens) [reply src/addons/meta/runtime.c:194]: That's fair, let's keep it this way for now then.

For ref, one way I've solved this in the past (different framework) is to compute the contiguous parts of a struct and memset/memcpy those:

```cpp
struct Foo {
  /* contiguous memset/memcpy-able block block */
  int foo;
  int bar;
  /* special member */
  char *str;
  /* contiguous memset/memcpy-able block block */
  int hello;
};
```
The code that generated the free routine would produce an array of instructions (similar to the array with callbacks you have, but with a bit more info) that had an instruction per contiguous block.

That's definitely more complex to implement, not suggesting we do that now.

File: src/addons/meta/runtime.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c:259]: Still, I am sharing `ecs_meta_rtt_ctx_t` for both structs and arrays and I need to add an entry for each hook type, so unfortunately I have to initialize each `ecs_meta_rtt_call_data_t`.
Anyway, I removed as much of the redundancy. Take a look now and see if that works.

[the remark this answers is not in the document]
Comment (SanderMertens) [reply src/addons/meta/runtime.c:259]: You're right, LGTM

File: src/addons/meta/runtime.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c:194]: Yes, this sounds like what I was proposing, that is, to "defragment" the memsets. Instead of having "instructions",  `ecs_meta_rtt_call_data_t` is prepared for this: point the `hook` member to a pseudo-ctor that just memsets but ignores `type_info`. This pseudo_ctor memsets everything and uses `count` as size.

```c
void flecs_meta_rtt_pseudo_ctor(
    void *ptr, 
    int32_t count, 
    const ecs_type_info_t *ti)
{
    (void) type_info;
    ecs_os_memset(ptr, 0, count);
}
```
Then, configure `ecs_meta_rtt_call_data_t` instance to have in `count` the sum of all sizes to memset.
Note that this is sort of a "private" hook, part of the call list and thus it is not directly callable from outside; nobody else would be able to call it with a bad `count` or anything else. So it looks like a hack, but it simplifies `ecs_meta_rtt_call_data_t` and the generic hook that has to execute the call list: It won't have to interpret "instructions", just blindly execute the call list, which would be faster I think.

The same applies to move and copy hooks, and that is why I'm doing:
```c
move_data->count = m->size * m->count;
move_data->hook.move = flecs_meta_rtt_default_move;
```

`flecs_meta_rtt_default_move` being:

```c
static void flecs_meta_rtt_default_move(void *dst_ptr, void *src_ptr, int32_t count, const ecs_type_info_t *type_info) {
  (void) type_info;  // silence unused warning
  ecs_os_memcpy(dst_ptr, src_ptr, count);
}
```

A second pass (future implementation) would then merge all call datas with `flecs_meta_rtt_default_move` into a big one with `count` equal to the sum of all sizes.

File: src/addons/meta/runtime.c [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+    }
+    if (move_hook_required) {
+      ecs_meta_rtt_call_data_t *move_data = ecs_vec_append_t(NULL, &rtt_ctx->moves, ecs_meta_rtt_call_data_t);
+      move_data->offset = m->offset;
+      move_data->type_info = member_ti;
+      if (member_ti->hooks.move) {
+        move_data->count = m->count;
+        move_data->hook.move = member_ti->hooks.move;
+      } else {
+        move_data->count = m->size * m->count;
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c:220]: This works like this in preparation for merging contiguous default move/copy areas of a struct. See explanation in https://github.com/SanderMertens/flecs/pull/1330#discussion_r1744125598

File: src/world.c [hunk 6]
```diff
@@ -1130,7 +1130,7 @@ void flecs_default_move_ctor_w_dtor(void *dst_ptr, void *src_ptr,
 }
 
 static
-void flecs_default_move(void *dst_ptr, void *src_ptr,
+void flecs_default_move_dtor(void *dst_ptr, void *src_ptr,
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/world.c]: reverted this rename.

File: src/addons/meta/runtime.c [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+      hooks.dtor = flecs_meta_rtt_dtor;
+    }
+    if (move) {
+      hooks.move = flecs_meta_rtt_move;
+    }
+    if (copy) {
+      hooks.copy = flecs_meta_rtt_copy;
+    }
+  } else {
+    hooks.lifecycle_ctx = NULL;
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c:153]: I have to set it to `NULL` because this could be the case that the hooks are being modified from requiring hooks to not requiring hooks. In this case I can't leave `lifecycle_ctx` or `lifecycle_ctx_free` function pointing to something, since there is no more `lifecycle_ctx`. That would leave the hooks in an inconsistent state.

That said, I modified it a bit to also clear `lifecycle_ctx_free`.

File: src/addons/meta/runtime.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c:179]: It is possible, but I'd lose the possibility to merge contiguous `flecs_default_ctor` in a struct. It is a great flag indicating the user just wants a memset(0). Since it is public, `FLECS_API`, we could document that the call may be merged together with others, but honestly I think this change is transparent to the user.

A struct like this:
```c
struct Person {
    char* name; // ecs_string_t
    char* last_name; // ecs_string_t
    char* city; // ecs_string_t
};
```
has 3 fields, each would call `flecs_default_ctor` for `ecs_string_t`. The proposed change detects this is the case and sets `flecs_default_ctor` at the struct level, instead of a call list with 3 items pointing to independent `flecs_default_ctor`s.

Additionally, I cannot distinguish between the author of a type not defining a ctor or explicitly setting `flecs_default_ctor`. By the time my code runs, I've lost that distinction; I only see `flecs_default_ctor`.

If the user wants to opt out of `flecs_default_ctor` merging, they can make their own ctor, but what would be the point?

File: src/addons/meta/runtime.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (SanderMertens) [reply src/addons/meta/runtime.c:153]: > because this could be the case that the hooks are being modified from requiring hooks to not requiring hooks

What is a scenario in which that could happen?

If it is possible that `lifecycle_ctx` is already set to something, wouldn't setting the value to `NULL` without freeing it result in a memory leak?

File: src/addons/meta/runtime.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (SanderMertens) [reply src/addons/meta/runtime.c:179]: It's more a principle of good API design, it should not override things that a user sets explicitly. If I do this:
```cpp
ecs_set_hooks(world, T, {
  .ctor = flecs_default_ctor
});
```
... then I expect that unconditionally, this always holds true:
```cpp
assert(ecs_get_type_info(world, T)->hooks.ctor == flecs_default_ctor);
```
It's true that this means you can't "collapse" multiple `flecs_default_ctor`'s, but that's a performance optimization that can be addressed in a different way in the future that doesn't have observable side effects for the user.

File: src/addons/meta/runtime.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c:153]: I don't know what scenario, but does not hurt to be clean, set the pointer to NULL after freeing, otherwise an attempt to free would happen again on world fini.

> If it is possible that lifecycle_ctx is already set to something, wouldn't setting the value to NULL without freeing it result in a memory leak?

It is freed above. Note that every time a member is added to a struct, the context is freed and then the call list is regenerated from scratch, like TypeSerializer.

> Can you add a test that fails when this value is not set to NULL?

I'll look into this, but I belive such test is together already with the tests of binding_ctx and ctx.

File: src/addons/meta/runtime.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (SanderMertens) [reply src/addons/meta/runtime.c:198]: I'm not a fan of "abusing" `count` as a `size` member. An invariant of the hook interface is that the memory pointed to by `dst_ptr` and `src_ptr` is `count * type_info->size`.

In a future performance optimization that collapses multiple fields into a single memset/memcpy, you could simply leave the callback to `NULL`, and call memset/memcpy directly. This is also better performance, since you don't have the overhead of calling a function pointer.

File: src/addons/meta/runtime.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (SanderMertens) [reply src/addons/meta/runtime.c:153]: > Note that every time a member is added to a struct, the context is freed and then the call list is regenerated

Gotcha, that answers my question :)

File: src/addons/meta/runtime.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c:179]: > `assert(ecs_get_type_info(world, T)->hooks.ctor == flecs_default_ctor);`

This holds true. I am not replacing your hook, because `T` is a component, meaning `EcsType.existing = true` and the new code does not run. The full example would be:

```c
  struct T {
    int value;
  };
  ECS_COMPONENT(world, T);
  ecs_set_hooks(world, T, {.ctor = flecs_default_ctor});
```

Another case would be that `T` is a field of a rtt struct. In that case, I just "consider" that you set T's ctor to `flecs_default_ctor` and will (in the future) create a collapsed call data, but I am not modifying the original hook. The optimization already in place is that if all fields in a struct happen to have `flecs_default_ctor` or `NULL`, then the resulting rtt struct's ctor will be `flecs_default_ctor` directly and we won't have a call per field and we won't have a `lifecycle_ctx` at all. That is actually the behavior prior to this PR.

```c
  struct T {
    int value;
  };
  ecs_entity_t t = ECS_COMPONENT(world, T);
  ecs_set_hooks(world, T, {.ctor = flecs_default_ctor});

  ecs_entity_t rtt_struct = ecs_struct(world, {.entity = ecs_entity(world, {.name = "RttStruct"}),
                                   .members = {{.name = "x", .type = t},
                                               {.name = "y", .type = ecs_id(ecs_u32_t)}});

  assert(ecs_get_type_info(world, T)->hooks.ctor == flecs_default_ctor); // holds true
  assert(ecs_get_type_info(world, rtt_struct)->hooks.ctor == flecs_default_ctor); // holds true

```

File: src/addons/meta/runtime.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c]: fixed

File: test/meta/project.json [hunk 8]
```diff
@@ -94,6 +94,22 @@
...
+                "move",
+                "copy",
+                "combined",
+                "trivial_array",
+                "array_ctor",
+                "array_dtor",
+                "array_move",
+                "array_copy",
+                "mixed"
+            ]
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply test/meta/project.json:115]: ### Tests with strings:
I believe these tests would be out of scope for RTT. `ecs_string_t` is not treated by RTT in any special way. From its point of view it is just a type that happens to have some hooks. RTT's responsibility is to make sure the underlying type's hooks are called when necessary, regardless of what type. To that effect, there is a mock type in the test cases that simulates something that holds a resource like an `ecs_string_t`. The mock allows to prove the hooks were called when they are supposed to, that's all RTT's tests care about. In the case of `ecs_string_t` whether its hooks actually do their job it is the responsibility of `ecs_string_t`'s unit tests.

### Tests with opaque types:

"root" opaque types: Prior to becoming opaques, these have `EcsType.existing == true`, defined via `ECS_COMPONENT`, and therefore get only a `flecs_default_ctor` by default (or keep whatever is set) and is not affected by this new code. There is a test that proves opaques are basically ignored.

Opaque types as RTT struct members, array and vector underlying types: Again, these are just types that happen to have hooks (or not). RTT code does not look at the `EcsType` of underlying types, it just cares about that type having lifecycle hooks or not. The mock in the tests simulate these hooks.

Therefore, I could say that a type being opaque or not does not affect lifecycle, only serialization.

### Tests of vectors of arrays of structs of nested structs with arrays of vectors of strings inside a struct with vectors 😅:
This would be a lot of fun, but the reality is that a RTT array, RTT vector or RTT struct itself becomes a type that has some hooks. Like a LEGO brick you can combine them however you like. Testing is focused on making sure the right hook is called when necessary, otherwise the list of test cases is infinite and does not really prove anything.
That said, the test case "mixed" tests an array of a struct that has arrays of structs, vectors of structs and nested structs, just for fun 🙂.

File: src/addons/meta/runtime.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c:214]: fixed

File: test/meta/project.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (SanderMertens) [reply test/meta/project.json:115]: > Like a LEGO brick you can combine them however you like. Testing is focused on making sure the right hook is called when necessary, otherwise the list of test cases is infinite and does not really prove anything.

These tests would prove that you _can_ combine things as lego blocks without something breaking. Even if it doesn't exercise different code paths today, it might after a refactor that has explicit behavior for vectors/strings/... (like what we discussed on discord).

These tests aren't for fun, they are to prevent future regressions :)

**Review by SanderMertens — CHANGES_REQUESTED** [review]

File: src/world.c [hunk 9]
```diff
@@ -6,178 +6,177 @@
...
-const ecs_entity_t EcsFlecsInternals =              FLECS_HI_COMPONENT_ID + 3;
-const ecs_entity_t EcsModule =                      FLECS_HI_COMPONENT_ID + 4;
-const ecs_entity_t EcsPrivate =                     FLECS_HI_COMPONENT_ID + 5;
-const ecs_entity_t EcsPrefab =                      FLECS_HI_COMPONENT_ID + 6;
-const ecs_entity_t EcsDisabled =                    FLECS_HI_COMPONENT_ID + 7;
-const ecs_entity_t EcsNotQueryable =                FLECS_HI_COMPONENT_ID + 8;
-
-const ecs_entity_t EcsSlotOf =                      FLECS_HI_COMPONENT_ID + 9;
-const ecs_entity_t EcsFlag =                        FLECS_HI_COMPONENT_ID + 10;
+const ecs_entity_t EcsWorld = FLECS_HI_COMPONENT_ID + 0;
```
Comment (SanderMertens) [inline src/world.c:24]: Restore original formatting

File: src/addons/meta/runtime.c [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+ * @file addons/meta/runtime.c
+ * @brief Runtime components.
+ */
+
+#include "flecs.h"
+#include "meta.h"
+
+#ifdef FLECS_META
+
+// Stores all the information necessary to forward a hook call to a nested/underlying type
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c]: fixed

File: src/addons/meta/runtime.c [hunk 11]
```diff
@@ -0,0 +1,328 @@
+/**
+ * @file addons/meta/runtime.c
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c]: fixed

File: src/addons/meta/runtime.c [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+ */
+
+#include "flecs.h"
+#include "meta.h"
+
+#ifdef FLECS_META
+
+// Stores all the information necessary to forward a hook call to a nested/underlying type
+typedef struct ecs_meta_rtt_call_data_t {
+  union {
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c]: fixed

File: src/addons/meta/runtime.c [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+// Lifecycle context for runtime types
+typedef struct ecs_meta_rtt_ctx_t {
+  ecs_vec_t ctors; /* vector<ecs_meta_rtt_call_data_t> */
+  ecs_vec_t dtors; /* vector<ecs_meta_rtt_call_data_t> */
+  ecs_vec_t copys; /* vector<ecs_meta_rtt_call_data_t> */
+  ecs_vec_t moves; /* vector<ecs_meta_rtt_call_data_t> */
+} ecs_meta_rtt_ctx_t;
+
+// Invokes underlying type's constructor/destructor using saved information in the lifecycle context
+static void flecs_meta_rtt_xtor(ecs_vec_t *xtor_data_vec, void *ptr, int32_t count, const ecs_type_info_t *type_info) {
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c]: fixed

File: src/addons/meta/runtime.c [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+  int cb_count = ecs_vec_count(xtor_data_vec);
+  for (int j = 0; j < count; j++) {
+    void *elem_ptr = ECS_ELEM(ptr, type_info->size, j);
+    for (int i = 0; i < cb_count; i++) {
+      ecs_meta_rtt_call_data_t *xtor_data = ecs_vec_get_t(xtor_data_vec, ecs_meta_rtt_call_data_t, i);
+      xtor_data->hook.xtor(ECS_OFFSET(elem_ptr, xtor_data->offset), xtor_data->count, xtor_data->type_info);
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c]: fixed

File: src/addons/meta/runtime.c [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+    }
+  }
+}
+
+// Generic constructor. It will read hook information call data from the type's lifecycle context and call
+// the constructors configured when the type was created.
+static void flecs_meta_rtt_ctor(void *ptr, int32_t count, const ecs_type_info_t *type_info) {
+  ecs_meta_rtt_ctx_t *rtt_ctx = (ecs_meta_rtt_ctx_t *) type_info->hooks.lifecycle_ctx;
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c]: fixed

File: src/addons/meta/runtime.c [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+  ecs_vec_t ctors; /* vector<ecs_meta_rtt_call_data_t> */
+  ecs_vec_t dtors; /* vector<ecs_meta_rtt_call_data_t> */
+  ecs_vec_t copys; /* vector<ecs_meta_rtt_call_data_t> */
+  ecs_vec_t moves; /* vector<ecs_meta_rtt_call_data_t> */
+} ecs_meta_rtt_ctx_t;
+
+// Invokes underlying type's constructor/destructor using saved information in the lifecycle context
+static void flecs_meta_rtt_xtor(ecs_vec_t *xtor_data_vec, void *ptr, int32_t count, const ecs_type_info_t *type_info) {
+  int cb_count = ecs_vec_count(xtor_data_vec);
+  for (int j = 0; j < count; j++) {
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c]: fixed

File: src/addons/meta/runtime.c [hunk 17]
```diff
@@ -0,0 +1,328 @@
...
+typedef struct ecs_meta_rtt_call_data_t {
+  union {
+    ecs_xtor_t xtor;
+    ecs_move_t move;
+    ecs_copy_t copy;
+  } hook;
+  const ecs_type_info_t *type_info;
+  int32_t offset;
+  int32_t count;
+} ecs_meta_rtt_call_data_t;
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c]: fixed

File: src/addons/meta/runtime.c [hunk 18]
```diff
@@ -0,0 +1,328 @@
...
+    move_hook_required |= member_ti->hooks.move != NULL;
+    copy_hook_required |= member_ti->hooks.copy != NULL;
+  }
+
+  // If any hook is required, then create a lifecycle context and configure a generic hook that will interpret that
+  // context:
+  ecs_meta_rtt_ctx_t *rtt_ctx = flecs_meta_rtt_configure_hooks(world, ti, ctor_hook_required, dtor_hook_required,
+                                                               move_hook_required, copy_hook_required);
+
+  if (!rtt_ctx)
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c]: fixed

File: src/addons/meta/runtime.c [hunk 19]
```diff
@@ -0,0 +1,328 @@
...
+    return;  // no hooks required
+
+  // At least a hook was configured, therefore examine each struct member to build the vector of calls that will then
+  // be executed by the generic hook handler:
+  for (int i = 0; i < ecs_vec_count(&struct_info->members); i++) {
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply src/addons/meta/runtime.c]: fixed

File: test/meta/src/RuntimeTypes.c [hunk 20]
```diff
@@ -0,0 +1,940 @@
...
+// Trivial structs should get flecs default constructor and nothing else
+void RuntimeTypes_trivial_struct(void) {
+  ecs_world_t *world = ecs_init();
+
+  // Create a component to use as subtype in the trivial struct
+  // Create a new component:
+  ecs_entity_t subtype = ecs_entity(world, {.name = "Some Subtype"});
+
+  // Configure as a struct:
+  ecs_struct(world, {.entity = subtype,
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply test/meta/src/RuntimeTypes.c]: fixed

File: test/meta/src/RuntimeTypes.c [hunk 21]
```diff
@@ -0,0 +1,940 @@
...
+                                               {.name = "s1", .type = nested_struct},
+                                               {.name = "y", .type = ecs_id(ecs_u32_t)},
+                                               {.name = "s2", .type = nested_struct}}});
+  return ecs_get_type_info(world, test_struct);
+}
+
+// Tests that a constructor is generated for a struct if at least a member has itself a constructor
+// Also tests if the generated constructor works.
+void RuntimeTypes_ctor(void) {
+  reset_test_ctor();
```
[the remark this answers is not in the document]
Comment (jpeletier) [reply test/meta/src/RuntimeTypes.c]: fixed

File: test/meta/project.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (jpeletier) [reply test/meta/project.json:114]: To solve this I wrote an "universal tester". You just define the structs, however complex you like, and a query picks them up and runs the tests on them. For exampe:

```cpp
    ecs_entity_t nightmare_struct = ecs_struct(
        world,
        {.entity = test_entity(world, 0, "nightmare_structº"),
         .members = {
             {.name = "a", .type = vector_of_strings},
             {.name = "b", .type = ecs_id(ecs_i32_t)},
             {.name = "c",
              .type = ecs_struct(
                  world,
                  {.members = {
                       {.name = "a",
                        .type = ecs_struct(
                            world, {.members = {{.name = "a",
                                                 .type = array_of_strings},
                                                {.name = "b",
                                                 .type = ecs_id(ecs_i32_t)}}})},
                       {.name = "b",
                        .type = ecs_struct(
                            world,
                            {.members =
                                 {{.name = "a", .type = vector_of_ints},
                                  {.name = "b", .type = ecs_id(ecs_i32_t)},
                                  {.name = "c",
                                   .type = ecs_struct(
                                       world,
                                       {.members =
                                            {{.name = "a",
                                              .type = vector_of_ints},
                                             {.name = "b",
                                              .type = ecs_id(ecs_i32_t)},
                                             {.name = "c",
                                              .type = vector_of_ints}}})}}})},
                       {.name = "c", .type = vector_of_strings}}})}}});
```

The good thing is that you can easily add any further types you want tested.

[the remark this answers is not in the document]
Comment (jpeletier) [reply test/meta/project.json:114]: > Can you add tests for:
> 
> * struct with strings
> * nested struct with strings
> * struct with array of strings
> [...]

Added these as entries in the universal tester.

**Comment by SanderMertens** [comment]
Thanks for the PR!
